docs/syntax_and_semantics/the_program.md

```diff
@@ -1,98 +1,80 @@
-# The Program
+# Program
 
-The program is a global object in which you can define types, methods and file-local variables.
+The program is the entirety of the source code worked by the compiler. The source gets parsed and compiled to an executable version of the program.
 
-```crystal
-# Defines a method in the program
-def add(x, y)
-  x + y
-end
+The program’s source code must be encoded in UTF-8.
 
-# Invokes the add method in the program
-add(1, 2) # => 3
-```
 
-A method's value is the value of its last expression; there's no need for explicit `return` expressions. However, explicit `return` expressions are possible:
 
-```crystal
-def even?(num)
-  if num % 2 == 0
-    return true
-  end
+## Top-level scope
 
-  return false
-end
-```
-
-When invoking a method without a receiver, like `add(1, 2)`, it will be searched for in the program if not found in the current type or any of its ancestors.
+Features such as types, constants, macros and methods defined outside any other namespace are in the top-level scope.
 
 ```crystal
+# Defines a method in the top-level scope
 def add(x, y)
   x + y
 end
 
-class Foo
-  def bar
-    # invokes the program's add method
-    add(1, 2)
+# Invokes the add method on the top-level scope
+add(1, 2) # => 3
+```
 
-    # invokes Foo's baz method
-    baz(1, 2)
-  end
+Local variables in the top-level scope are file-local and not visible inside method bodies.
 
-  def baz(x, y)
-    x * y
-  end
+```crystal
+x = 1
+
+def add(y)
+  x + y # error: undefined local variable or method 'x'
 end
+
+add(2)
 ```
 
-If you want to invoke the program's method, even though the current type defines a method with the same name, prefix the call with `::`:
+Private features are also only visible in the current file.
+
+A double colon prefix (`::`) unambiguously references a namespace, constant, method or macro in the top-level scope:
 
 ```crystal
-def baz(x, y)
-  x + y
+def baz
+  puts "::baz"
 end
 
-class Foo
-  def bar
-    baz(4, 2)   # => 2
-    ::baz(4, 2) # => 6
-  end
+CONST = "::CONST"
 
-  def baz(x, y)
-    x - y
+module A
+  def self.baz
+    puts "A.baz"
   end
-end
-```
 
-Variables declared in a program are not visible inside methods:
+  # Without prefix, resolves to the method in the local scope
+  baz
 
-```crystal
-x = 1
+  # With :: prefix, resolves to the method in the top-level scope
+  ::baz
 
-def add(y)
-  x + y # error: undefined local variable or method 'x'
-end
+  CONST = "A::Const"
 
-add(2)
+  p! CONST   # => "A::CONST"
+  p! ::CONST # => "::CONST"
+end
 ```
 
```

Comment:
How about adding:
```suggestion
!!! tip
   You can check the list of methods and constants included accessible from the top-level in the [API documentation](https://crystal-lang.org/api/toplevel.html).

```

Reply:
Yeah, I thought about that earlier as well. But it's confusing. The API docs show only what's in the standard library. Most typical program have a lot more stuff.
Perhaps this could be rephrased somehow, but I'm not sure if such a reference even makes sense here at all.

Reply:
Yeah, today I don't like this either 😅